hazelcast/src/hazelcast/client/cluster/impl/VectorClock.cpp

```diff
@@ -0,0 +1,83 @@
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <boost/foreach.hpp>
+#include <set>
+
+#include "hazelcast/client/cluster/impl/VectorClock.h"
+#include "hazelcast/client/cluster/impl/ClusterDataSerializerHook.h"
+#include "hazelcast/client/serialization/ObjectDataInput.h"
+
+namespace hazelcast {
+    namespace client {
+        namespace cluster {
+            namespace impl {
+
+                VectorClock::VectorClock() {}
+
+                typedef std::vector<std::pair<std::string, boost::shared_ptr<int64_t> > > TimestampVector;
+
+                std::vector<std::pair<std::string, int64_t> > VectorClock::entrySet() {
+                    TimestampVector entries = replicaTimestamps.entrySet();
+                    std::vector<std::pair<std::string, int64_t> > result;
+                    BOOST_FOREACH(const TimestampVector::value_type &entry , entries) {
+                        result.push_back(std::make_pair(entry.first, *entry.second));
+                    }
+
+                    return result;
+                }
+
+                void VectorClock::setReplicaTimestamp(const std::string &replicaId, int64_t timestamp) {
+                    replicaTimestamps.put(replicaId, boost::shared_ptr<int64_t >(new int64_t(timestamp)));
+                }
+
+                bool VectorClock::isAfter(VectorClock &other) {
+                    bool anyTimestampGreater = false;
+                    BOOST_FOREACH (const TimestampVector::value_type &otherEntry , other.replicaTimestamps.entrySet()) {
+                        const std::string &replicaId = otherEntry.first;
+                        const boost::shared_ptr<int64_t> &otherReplicaTimestamp = otherEntry.second;
+                        boost::shared_ptr<int64_t> localReplicaTimestamp = getTimestampForReplica(replicaId);
+
+                        if (localReplicaTimestamp.get() == NULL || localReplicaTimestamp < otherReplicaTimestamp) {
+                            return false;
+                        } else if (*localReplicaTimestamp > *otherReplicaTimestamp) {
+                            anyTimestampGreater = true;
+                        }
+                    }
+                    // there is at least one local timestamp greater or local vector clock has additional timestamps
+                    return anyTimestampGreater || !containsAll(other.replicaTimestamps.keys(), replicaTimestamps.keys());
```

Comment:
if you reach here it is guaranteed that `replicaTimstamps` has all the keys that `other.replicaTimestamps` has. You can simply check if number of keys in `replicaTimestamps` is greater than the number of keys in `other.replicaTimestamps` instead of `containsAll`

Reply:
Yes, that can be done. But should we do this for Java client first?: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/cluster/impl/VectorClock.java#L104

We can also break the loop at line 56 where the anyTimestampGreater is true.

Should we do these two first at Java and then at C++ client?

Reply:
@mustafaiman make sense. I will do this change on hava. 

@ihsandemir I think we can't break the loop. For example following test fails when we break the loop at 56
assertFalse(vectorClock("A", 2).isAfter(vectorClock("A", 1, "B", 1)));

Reply:
https://github.com/hazelcast/hazelcast/pull/13477

Reply:
Yes, we can't break the loop as those two vector clocks are concurrent, neither is greater than the other.